action/governance/withdrawFunds.go

```diff
@@ -0,0 +1,234 @@
+}
+
+func runWithdraw(ctx *action.Context, signedTx action.RawTx) (bool, action.Response) {
+	withdrawProposal := WithdrawFunds{}
+	err := withdrawProposal.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, action.Response{}
+	}
+
+	// 1. Check if Proposal already exists, if so, check the withdraw requirement:
+	//    a. the funding goal is not reached
+	//    b. the funding height is reached
+
+	proposal, _, err := ctx.ProposalMasterStore.Proposal.QueryAllStores(withdrawProposal.ProposalID)
+	if err != nil {
+		ctx.Logger.Error("Proposal does not exist :", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_does_not_exist"),
+		}
+		return false, result
+	}
+	currentFundsForProposal := governance.GetCurrentFunds(proposal.ProposalID, ctx.ProposalMasterStore.ProposalFund)
+	// if funding goal is reached or there is still time for funding
+	if currentFundsForProposal.BigInt().Cmp(proposal.FundingGoal.BigInt()) >= 0 || ctx.Header.Height <= proposal.FundingDeadline {
+		ctx.Logger.Error("Proposal does not meet withdraw requirement", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_does_not_meet_withdraw_requirement"),
+		}
+		return false, result
+	}
+	// 2. change outcome, status, state
+	proposal.Outcome = governance.ProposalOutcomeInsufficientFunds
+	proposal.Status = governance.ProposalStatusCompleted
+	err = ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateFailed).Set(proposal)
+	if err != nil {
+		ctx.Logger.Error("Failed to add proposal to FAILED store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_add_proposal_to_failed_store"),
+		}
+		return false, result
+	}
+	ok, err := ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateActive).Delete(proposal.ProposalID)
+	if err != nil || !ok {
+		ctx.Logger.Error("Failed to delete proposal from ACTIVE store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_delete_proposal_from_active_store"),
+		}
+		return false, result
+	}
+
+	// 3. Check if the contributor has funded this proposal, if so, get the amount of funds
+	proposalFund, err := governance.GetCurrentFundsByContributor(proposal.ProposalID, withdrawProposal.Contributor, ctx.ProposalMasterStore.ProposalFund)
+	if err != nil {
+		ctx.Logger.Error("No available funds to withdraw for this contributor :", withdrawProposal.Contributor)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "no_available__fund_to_withdraw_for_this_contributor"),
+		}
+		return false, result
+	}
+
+	// 4. Check if the contributor has sufficient funds to withdraw for that proposal
+	withdrawAmount := balance.NewAmountFromBigInt(withdrawProposal.WithdrawValue.Value.BigInt())
```

Comment:
The DeductFunds will do this check inside, right?

Reply:
yeah, I'll remove this and test if this is covered